Rate limiting stale sst files' deletion during recovery

jowlyzhang opened the pull request:
As titled. If SstFileManager is available, deleting stale sst files will be delegated to it so it can be rate limited.

hx235 commented:
@jowlyzhang This seems like a feature worth a UT - either a new UT or a modified one. Could you add one?

hx235 reviewed (changes requested):
@jowlyzhang This seems like a feature worth a unit test - either a new UT or a modified one. Could you add one?

jowlyzhang commented:
> @jowlyzhang This seems like a feature worth a unit test - either a new UT or a modified one. Could you add one?

`DBBasicTest.LastSstFileNotInManifest` should cover this code path. It would fail with the deletion logic commented out. (I also need to comment out some other backup deletion logic to make it fail). But I think this change itself is not risky or introducing completely new behavior. It's just changing from using `env_->DeleteFile` to the commonly used deletion API `DeleteDBFile` so that rate limiting is exercised if it's available through SstFileManager.

hx235 commented:
> It's just changing from using env_->DeleteFile to the commonly used deletion API DeleteDBFile so that rate limiting is exercised if it's available through SstFileManager.

I'm a bit worried about if there is already a test case for stale SST files's deletion **actually** going through rate limiting upon **recovery**. 

`DBBasicTest.LastSstFileNotInManifest` did go through your change on delegation but the SstFileManager is not set to rate limiting anything so no rate limiting upon recovery's file deletion.

Surprisingly db_sst_test.cc has a few SstFileManager with rate limit configured but it does not go through your `DeleteDBFile()` change . The rate-limiting shows effect in BG flush/compaction trying to clean up stuff instead of in recovery. BG flush/compaction are paths different enough from recovery.

Because of the above, it's not obvious to me what is the existing UT that actually covers the path this PR is aiming for. Ideally there should be one to help catch unintentional change on recovery path that ends up disabling rate limiting.

Are you already aware of such UT? I don't find this UT very obvious so if there is already one, noting that in PR summary will be greatly appreciated :) 

If there isn't one, maybe the one db_sst_test.cc is worth modifying.

hx235 commented:
@jowlyzhang 

I'm a bit worried about whether there already exists a test case for stale SST files's deletion **actually** going through rate limiting upon **recovery**.

`DBBasicTest.LastSstFileNotInManifest` did go through your change on delegation but the SstFileManager is not set to rate limiting anything so no rate limiting upon recovery's file deletion.

Surprisingly `db_sst_test.cc` has a few SstFileManager with rate limit configured but it does not go through your DeleteDBFile() change . The rate-limiting shows effect in BG flush/compaction trying to clean up stuff instead of in recovery. BG flush/compaction are paths different enough from recovery.

`TEST_F(BlobDBTest, SstFileManagerRestart)` looks like a close one but it does not go through you DeleteDBFile() change, probably due to BlobDB does not go through the same path for delete files upon recovery?

Because of the above, it's not obvious to me what is the existing UT that actually covers the path this PR is aiming for. Ideally there should be one to help catch unintentional change on recovery path that ends up disabling rate limiting, even though the current PR is not risky.

Are you already aware of such UT? I don't find this UT very obvious so if there is already one, noting that in PR summary will be greatly appreciated :)

edit: If there isn't one, maybe the one db_sst_test.cc is worth modifying.

jowlyzhang commented:
> @jowlyzhang
> 
> I'm a bit worried about whether there already exists a test case for stale SST files's deletion **actually** going through rate limiting upon **recovery**.
> 
> `DBBasicTest.LastSstFileNotInManifest` did go through your change on delegation but the SstFileManager is not set to rate limiting anything so no rate limiting upon recovery's file deletion.
> 
> Surprisingly `db_sst_test.cc` has a few SstFileManager with rate limit configured but it does not go through your DeleteDBFile() change . The rate-limiting shows effect in BG flush/compaction trying to clean up stuff instead of in recovery. BG flush/compaction are paths different enough from recovery.
> 
> `TEST_F(BlobDBTest, SstFileManagerRestart)` looks like a close one but it does not go through you DeleteDBFile() change, probably due to BlobDB does not go through the same path for delete files upon recovery?
> 
> Because of the above, it's not obvious to me what is the existing UT that actually covers the path this PR is aiming for. Ideally there should be one to help catch unintentional change on recovery path that ends up disabling rate limiting, even though the current PR is not risky.
> 
> Are you already aware of such UT? I didn't find this UT very obvious so if there is already one, noting that in PR summary will be greatly appreciated :)
> 
> edit: If there isn't one, maybe the one db_sst_test.cc is worth modifying.

I see what you mean. Having rate limiting is important for production, it's indeed very useful to have it covered in unit tests to prevent regression. Thank you for finding the test coverage in `db_sst_test`. I modified an existing test covering cleaning up trash files to also cover deleting obsolete sst files for recovery and added rate limiting related assertions for them.

hx235 commented on db/db_sst_test.cc:
```diff
@@ -937,12 +937,21 @@ INSTANTIATE_TEST_CASE_P(DBWALTestWithParam, DBWALTestWithParam,
+  sfm->SetDeleteRateBytesPerSecond(1024 * 1024);
+  sfm->delete_scheduler()->SetMaxTrashDBRatio(1000.0);
+
+  int bg_delete_file = 0;
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+      "DeleteScheduler::DeleteTrashFile:DeleteFile",
+      [&](void* /*arg*/) { bg_delete_file++; });
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
```
Would usage of stats like FILES_DELETED_FROM_TRASH_QUEUE/FILES_DELETED_IMMEDIATELY replace the usage of sync points?

 We'd like to reduce usage of sync point in test whenever possible since they make refactoring existing codes more difficult.

jowlyzhang replied:
That's good point. I will check this in a follow up PR to clean up them all since quite a lot of tests in this file use this mechanism.

hx235 reviewed (approved):
LGTM with one comment on test. Also maybe mention this change in "Behavior Changes" since users with rate limited SstManager will experience slower db open (for the good)
